grep-XXXX-oots-same-as-modules.md

```diff
@@ -0,0 +1,103 @@
+# GREP [XXXX] -- OOTs at same level as modules
+
+[Note: All parts in square brackets are meant to be replaced as part of
+authoring a new GREP]
+
+- Original Author: Josh Morman <jmorman@gnuradio.org>
+- Status: Draft 
+
+History:
+- 09-Nov-2021: Initial Draft
+
+## Abstract
+
+Based off of [Issue #3006](https://github.com/gnuradio/gnuradio/issues/3006)
+
+New OOTs should taken on a structure that more closely resembles the in-tree 
+modules.  This allows them to be more coupled with gnuradio, rather than living 
+in their own non-GR associated directories and python modules.  We consider 3
+aspects to improving the OOT structure
+
+1) Python module under gnuradio
+2) C++ header under gnuradio
+3) CMake naming involving gnuradio
+
+## Copyright / License
+
+CC-BY-ND
+
+## Motivation
+
+When creating a GNU Radio OOT module using `gr-modtool`, this creates a python 
+module and a set of headers/libs which in naming bear no association with GNU Radio.
+
+If a module is a GNU Radio wrapping of some other well known module, there can
+be naming conflicts, or at the least confusion.  Take for example cuda: If I want
+to use gr-modtool and create a new module gr-cuda - now I have in my prefix
+
+```python
+import cuda
+```
+and
+```c++
+#include <cuda/my_cuda_block.h>
+```
+
+Both of these use the broad naming `cuda` for something that is intended to be 
+GNU Radio's wrapping of some cuda functionality - a broad overreach
+
+Another motivation is simply for commonality between the in-tree modules and OOTs
+and encourage OOT developers to easily deploy robust CMake module files that 
+name their modules similar to other GNU Radio components.  For example instead of
+`find_package(cuda)` - which is problematic, a user that is integrating elements
+of an OOT into their OOT should call `find_package(gnuradio-cuda)` or even possibly:
+`find_package(gnuradio COMPONENTS cuda)` as a stretch
+
+
+## Description
+
+### Required Changes
+1) Python module structure
+2) Header structure
+3) CMake structure
+
+Refer to example OOT in these examples as howto
+
+#### Python module structure
+- Python source goes from `gr-howto/python/` to `gr-howto/python/howto/`
+- Likewise for bindings directory
+- Python module installation path goes from `PYTHON_DIR/howto` to `PYTHON_DIR/gnuradio/howto`
+
+The net result of this is `from gnuradio import howto`
```

Comment:
But doesn't that require the `/usr/lib/python3/site-packages/gnuradio/__init__.py` to be aware of that `howto`, which might be installed in a different prefix? (my usecase here is: I use a site-wide GNU Radio installation, but have a `PYTHONPATH`, `PATH`, `LD_LIBRARY_PATH`-prefix in my home directory so that for OOT development/testing I don't need to sudo install stuff)

Reply:
I remember some osmocom OOTs doing this in the past, I don't know whether/how that worked!

Reply:
Interesting.  The plot thickens.  Will have to do some research to figure out how it would work in this use case. 

The other use case I'd like to figure out is when running QA tests on the OOT module - can that be done without installing first.  Seems like the solution here would be for both

Reply:
I think this can be accomplished with `pkgutils`
```python
from pkgutil import extend_path
__path__ = extend_path(__path__, __name__)
```
Thus anything in the PYTHONPATH that has the name `gnuradio` will be considered as an extension of this module